src/consumer/consumerGroup.js

```diff
@@ -566,4 +614,65 @@ module.exports = class ConsumerGroup {
   hasSeekOffset({ topic, partition }) {
     return this.seekOffset.has(topic, partition)
   }
+
+  /**
+   * For each of the partitions find the best nodeId to read it from
+   *
+   * @param {string} topic
+   * @param {number[]} partitions
+   * @returns {{[nodeId: number]: number[]}} per-node assignment of partitions
+   * @see Cluster~findLeaderForPartitions
+   */
+  // Invariant: The resulting object has each partition referenced exactly once
+  findReadReplicaForPartitions(topic, partitions) {
+    const partitionMetadata = this.cluster.findTopicPartitionMetadata(topic)
+    const preferredReadReplicas = this.preferredReadReplicasPerTopicPartition[topic]
+    return partitions.reduce((result, id) => {
+      const partitionId = parseInt(id, 10)
+      const metadata = partitionMetadata.find(p => p.partitionId === partitionId)
+      if (!metadata) {
+        return result
+      }
+
+      if (metadata.leader === null || metadata.leader === undefined) {
+        throw new KafkaJSError('Invalid partition metadata', { topic, partitionId, metadata })
+      }
+
+      // Pick the preferred replica if there is one, and it isn't known to be offline, otherwise the leader.
+      let nodeId = metadata.leader
+      if (preferredReadReplicas) {
+        const { nodeId: preferredReadReplica, expireAt } = preferredReadReplicas[partitionId] || {}
+        if (Date.now() > expireAt) {
```

Comment:
This is a bit nitpicky, but given the name of `expireAt`, I would say that we shouldn't use an entry that is expiring right now.

```suggestion
        if (Date.now() >= expireAt) {
```